iconservice/fee/fee_engine.py

```diff
@@ -0,0 +1,844 @@
+
+        deposit_info = DepositInfo(score_address)
+
+        # Appends all deposits
+        for deposit in self._deposit_generator(context, deposit_meta.head_id):
+            deposit_info.deposits.append(deposit)
+
+            # Retrieves available virtual STEPs and deposits
+            if block_height < deposit.expires:
+                deposit_info.available_virtual_step += deposit.remaining_virtual_step
+                deposit_info.available_deposit += \
+                    max(deposit.remaining_deposit - deposit.min_remaining_deposit, 0)
+
+        return deposit_info
+
+    def add_deposit(self,
+                    context: 'IconScoreContext',
+                    tx_hash: bytes,
+                    sender: 'Address',
+                    score_address: 'Address',
+                    amount: int,
+                    block_height: int,
+                    term: int) -> None:
+        """
+        Deposits ICXs for the SCORE.
+        It may be issued the virtual STEPs for the SCORE to be able to pay share fees.
+
+        :param context: IconScoreContext
+        :param tx_hash: tx hash of the deposit transaction
+        :param sender: ICX sender
+        :param score_address: SCORE
+        :param amount: amount of ICXs in loop
+        :param block_height: current block height
+        :param term: deposit term in blocks
+        """
+        # [Sub Task]
+        # - Deposits ICX
+        # - Calculates Virtual Step
+        # - Updates Deposit Data
+
+        if not (self._MIN_DEPOSIT_AMOUNT <= amount <= self._MAX_DEPOSIT_AMOUNT):
+            raise InvalidRequestException('Invalid deposit amount')
+
+        if not (self._MIN_DEPOSIT_TERM <= term <= self._MAX_DEPOSIT_TERM):
+            raise InvalidRequestException('Invalid deposit term')
+
+        self._check_score_ownership(context, sender, score_address)
+
+        deposit_meta = self._get_or_create_deposit_meta(context, score_address)
+
+        # Withdraws from sender's account
+        sender_account = self._icx_storage.get_account(context, sender)
+        sender_account.withdraw(amount)
+        self._icx_storage.put_account(context, sender, sender_account)
+
+        deposit = Deposit(tx_hash, score_address, sender, amount)
+        deposit.created = block_height
+        deposit.expires = block_height + term
+        deposit.virtual_step_issued = \
+            self._calculate_virtual_step_issuance(amount, deposit.created, deposit.expires)
+        deposit.prev_id = deposit_meta.tail_id
+
+        self._insert_deposit(context, deposit)
+
+    def _insert_deposit(self, context, deposit):
+        """
+        Inserts deposit data to storage
+        """
+
+        deposit_meta = self._get_or_create_deposit_meta(context, deposit.score_address)
+
+        deposit.prev_id = deposit_meta.tail_id
+        self._fee_storage.put_deposit(context, deposit.id, deposit)
+
+        # Link to old last item
+        if deposit_meta.tail_id is not None:
```

Comment:
Conceptually, using `deposit.prev_id` is more natural here instead of using `deposit_meta.tail_id`.
```suggestion
        if deposit.prev_id is not None:
            prev_deposit = self._fee_storage.get_deposit(context, deposit.prev_id)
```